src/validator.js

```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+(function(exports) {
+    var Validator = function() {
+        this._numberOfFrames = 0;
+        this.frameBuilder = new FrameBuilder();
+    };
+
+    Validator.prototype = {
+        validate: function(rolls) {
+            return this._validateScores(rolls) && this._validateFrames(rolls);
+        },
+
+        resetFrames: function(rolls) {
+            this._numberOfFrames = 0;
+        },
+
+        _validateScores: function(rolls) {
+            for(let i = 0; i <= rolls.length; i++) {
+                if (rolls[i] > 10) {
+                    return false;
+                }
+            }
+            return true;
+        },
+
+        _validateFrames: function(rolls) {
+            var firstFrame;
+
+            if (this._numberOfFrames === 9) {
+                this._buildFrame(rolls, rolls, this._numberOfFrames);
+                return this._validateLastFrame(rolls)
+            }
+
+            if (this._isStrike(rolls)) {
+                firstFrame = firstRolls(rolls, 1);
+                this._buildFrame(firstFrame, firstRolls(rolls, 3));
+            } else if (this._isSpare(rolls)) {
+                firstFrame = firstRolls(rolls, 2);
+                this._buildFrame(firstFrame, firstRolls(rolls, 3));
+            } else {
+                firstFrame = firstRolls(rolls, 2);
+                this._buildFrame(firstFrame, firstRolls(rolls, 2));
+            }
+
+            return this._validateFrame(firstFrame) && this._validateFrames(removeRolls(rolls, firstFrame.length));
```

Comment:
In response to your question – "is there a way to make this clearer?", I think this is pretty clear. The recursion is neat, and I don't have a big problem with the `if/else` statement – each one is clearly handling a different kind of roll. It's pretty easy to read and understand.

Some quick wins – there are some magic numbers in here, and your code will be cleaner if you extract them. What is `1`, `3`, `2`? This might lead, in turn, to different versions of the `firstRolls` method.

You might also gain by thinking in a slightly more OO way. For instance – you have some ways of representing a frame that also knows about other rolls (`this._buildFrame(firstFrame, firstRolls(rolls, 3))`). Could these be special kinds of object? For instance:

```javascript
if (this._isStrike(rolls)) {
  this._buildStrikeFrame(rolls)
}
```

Then, in `buildStrikeFrame`:

```javascript
_buildStrikeFrame: function(rolls) {
  this.frameBuilder.buildStrikeFrame(rolls, this._numberOfFrames++)
}
```

And in `buildStrikeFrame`:

```javascript
buildStrikeFrame: function(allRolls, frameNumber) {
  var firstFrame = firstRolls(rolls, 1)
  var otherAffectedFrames = firstRolls(rolls, 3)
  return new Frame(firstFrame, otherAffectedFrames)
}
```

One rule-of-thumb I'm using here is 'if you are pulling information out of a data structure and passing that to an object, just pass the whole data structure to the object instead'. So just pass `rolls` to the `FrameBuilder` and let it figure out how to deal with them.